File: services/typescript/koa/src/routes/handlers/put.ts
import type { RouterContext } from "@koa/router";
import { type State, parseTask, parseId } from "app";
import { getLogger } from "telemetry";
import { AppErrors, sendError } from "../errors.js";

export function putHandler(state: State) {
  return async (ctx: RouterContext) => {
    const logger = getLogger();

    const id = parseId(ctx.params.id);
    if (!id.success) {
      logger.warn({ method: ctx.req.method, path: ctx.path }, "Invalid path");
      return sendError(ctx, AppErrors.InvalidPath);
    }

    const task = parseTask(ctx.request.body);
    if (!task.success) {
      logger.warn(
        { method: ctx.req.method, path: ctx.path },
        "Invalid body JSON"
      );
      return sendError(ctx, AppErrors.InvalidJsonBody);
    }

    const previous_task = state.tasks.get(id.data);

    if (previous_task === undefined) {
      logger.warn(
        { id: id.data, method: "PUT", path: ctx.path },
        "Task not found"
      );
      return sendError(ctx, AppErrors.TaskNotFound);
    }

    state.tasks.set(id.data, task.data);

    logger.info(
      {
        id: id.data,
        from_secret: previous_task.secret.length,
        to_secret: task.data.secret.length,
        from_operation: previous_task.operation.toString().toLowerCase(),
        to_operation: task.data.operation.toString().toLowerCase(),
        method: "PUT"
      },
      "Overwrote task"
    );

    ctx.status = 200;
    ctx.body = task.data;
  };
}

File: services/python/starlette/src/routes/handlers/put.py
import structlog
import json
from uuid import UUID
from pydantic import ValidationError
from starlette.requests import Request
from starlette.responses import Response, JSONResponse

from app.params import parse_task
from app.task import Task
from app.state import AppState
from routes.errors import send_error, AppErrors


async def put_handler(request: Request) -> Response:
    logger = structlog.get_logger()

    id: UUID = request.path_params["id"]

    try:
        body = await request.json()
        new_task: Task = parse_task(body)
    except (json.JSONDecodeError, ValidationError):
        logger.warn("Invalid body JSON", method=request.method, path=request.url.path)
        return send_error(AppErrors.InvalidJsonBody)

    state: AppState = request.app.state.app_state

    async with request.app.state.tasks_lock:
        task: Task | None = state.tasks.get(id)

        if task is None:
            logger.warn(
                "Task not found",
                id=str(id),
                method=request.method,
                path=request.url.path,
            )
            return send_error(AppErrors.TaskNotFound)

        state.tasks[id] = new_task

    logger.info(
        "Overwrote task",
        id=str(id),
        from_secret=len(task.secret),
        to_secret=len(new_task.secret),
        from_operation=task.operation.lower(),
        to_operation=new_task.operation.lower(),
        method=request.method,
    )

    return JSONResponse(new_task.model_dump())
